packages/tpetra/core/src/Tpetra_CrsMatrix_def.hpp

```diff
@@ -59,32 +59,42 @@
 #include "Tpetra_Details_computeOffsets.hpp"
 #include "Tpetra_Details_copyOffsets.hpp"
 #include "Tpetra_Details_createMirrorView.hpp"
-#include "Tpetra_Details_getDiagCopyWithoutOffsets.hpp"
 #include "Tpetra_Details_gathervPrint.hpp"
+#include "Tpetra_Details_getDiagCopyWithoutOffsets.hpp"
 #include "Tpetra_Details_leftScaleLocalCrsMatrix.hpp"
 #include "Tpetra_Details_Profiling.hpp"
 #include "Tpetra_Details_rightScaleLocalCrsMatrix.hpp"
-#include "Teuchos_SerialDenseMatrix.hpp"
 #include "KokkosSparse_getDiagCopy.hpp"
 #include "Tpetra_Details_copyConvert.hpp"
+#include "Tpetra_Details_iallreduce.hpp"
 #include "Tpetra_Details_getEntryOnHost.hpp"
 #include "Tpetra_Details_packCrsMatrix.hpp"
 #include "Tpetra_Details_unpackCrsMatrixAndCombine.hpp"
+#include "Teuchos_FancyOStream.hpp"
+#include "Teuchos_RCP.hpp"
+#include "Teuchos_SerialDenseMatrix.hpp" // unused here, could delete
 #include <memory>
 #include <sstream>
 #include <typeinfo>
 #include <vector>
 
-#include <Teuchos_RCP.hpp>
-
-#include <Teuchos_FancyOStream.hpp>
-
 using Teuchos::rcpFromRef;
 
 namespace Tpetra {
 
 namespace { // (anonymous)
 
+  std::shared_ptr< ::Tpetra::Details::CommRequest>
+  iallreduceIntRaw (const int& localValue,
+		    int& globalValue,
+		    const ::Teuchos::EReductionType op,
+		    const Teuchos::Comm<int>& comm)
+  {
+    Kokkos::View<const int*, Kokkos::HostSpace> localView (&localValue);
+    Kokkos::View<int*, Kokkos::HostSpace> globalView (&globalValue);
+    return ::Tpetra::Details::iallreduce (localView, globalView, op, comm);
```

Comment:
This has a nonblocking fall-back implementation in case `MPI_VERSION < 3`.  (It just defers the all-reduce until `wait()`.)